### PR TITLE
Created separate file for VariableMap

### PR DESCRIPTION
**Pull request by marisaleung**
This change is [Reviewable](https://reviewable.io/reviews/google/blockly/1079)

### CODE REVIEW

**Comment by rachel-fenichel** [comment]
Unit tests are failing, since you added a new file.

Please rebuild and add only blockly_uncompressed.js to this PR.

**Comment by marisaleung** [comment]
Done

File: core/variable_map.js [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+/**
+ * Delete a variable.
+ * @param {Blockly.VariableModel} variable Variable to delete.
+ */
+Blockly.VariableMap.prototype.deleteVariable = function(variable) {
+  var type = variable.type;
+  for (var i = 0, tempVar; tempVar = this.variableMap_[type][i]; i++) {
+    if (Blockly.Names.equals(tempVar.name, variable.name)) {
+      delete this.variableMap_[type][i];
```
Comment (rachel-fenichel) [inline core/variable_map.js]: I think you don't need to call delete--it should get handled by gc once it's removed from the list, and variableModel doesn't have any special disposal code.

File: core/variable_map.js [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+      return;
+    }
+  }
+};
+
+/**
+ * Find the variable by the given name and return it. Return null if it is not
+ *     found.
+ * @param {!string} name The name to check for.
+ * @return {?Blockly.VariableModel} the variable with the given name.
```
Comment (rachel-fenichel) [inline core/variable_map.js]: the variable with the given name, or null if it was not found.

File: core/variable_map.js [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+      if (variable.getId() == id) {
+        return variable;
+      }
+    }
+  }
+  return null;
+};
+
+/**
+ * Find the variable with the specified type. If type is null, return list of
```
Comment (rachel-fenichel) [inline core/variable_map.js]: "Get a list containing all of the variables of a specified type.  If type is null, ..."

**Comment by marisaleung** [comment]
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[core/variable_map.js, line 80 at r1](https://reviewable.io:443/reviews/google/blockly/1079#-KjFN727c16Ee6QrwXP1-r1-80:-KjFNKKod67tq7b2caJQ:b-896fix) ([raw file](https://github.com/google/blockly/blob/23526f44f79dfa75a3f6176d737fba6a3b6ba3d8/core/variable_map.js#L80)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Nit: Move the else if onto the previous line
</blockquote></details>

Done.

---

*[core/variable_map.js, line 115 at r1](https://reviewable.io:443/reviews/google/blockly/1079#-KjFN727c16Ee6QrwXP1-r1-115:-KjFNRQZGTLgOt0zTM4C:b-896fix) ([raw file](https://github.com/google/blockly/blob/23526f44f79dfa75a3f6176d737fba6a3b6ba3d8/core/variable_map.js#L115)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

This is the case where nothing went wrong but the variable already existed, correct?  If so, add a comment: "// The variable already exists and has the same id and type."
</blockquote></details>

Done.

---

*[core/variable_map.js, line 129 at r1](https://reviewable.io:443/reviews/google/blockly/1079#-KjFN727c16Ee6QrwXP1-r1-129:-KjFNuyxyLwmXnYAVC4A:b-896fix) ([raw file](https://github.com/google/blockly/blob/23526f44f79dfa75a3f6176d737fba6a3b6ba3d8/core/variable_map.js#L129)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

nit: `else` should share a line with the previous closing brace.  You can just put the comment inside here, or else remove it.
</blockquote></details>

Done.

---

*[core/variable_map.js, line 140 at r1](https://reviewable.io:443/reviews/google/blockly/1079#-KjFN727c16Ee6QrwXP1-r1-140:-KjFOoQEtly_Timp0xbu:b-896fix) ([raw file](https://github.com/google/blockly/blob/23526f44f79dfa75a3f6176d737fba6a3b6ba3d8/core/variable_map.js#L140)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Store this.variableMap_[type] in a variable outside of the loop for clarity.
</blockquote></details>

Done.

---

*[core/variable_map.js, line 141 at r1](https://reviewable.io:443/reviews/google/blockly/1079#-KjFN727c16Ee6QrwXP1-r1-141:-KjFPAmHhDgaMzpPdcgC:bxrxq1q) ([raw file](https://github.com/google/blockly/blob/23526f44f79dfa75a3f6176d737fba6a3b6ba3d8/core/variable_map.js#L141)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Why not check by id?  Then you don't have to depend on additional blockly code--it's just a string comparison.
</blockquote></details>

Good point.
Done.

---

*[core/variable_map.js, line 142 at r1](https://reviewable.io:443/reviews/google/blockly/1079#-KjFN727c16Ee6QrwXP1-r1-142:-KjFPcXKwtEzejsuCYmk:bkplsqw) ([raw file](https://github.com/google/blockly/blob/23526f44f79dfa75a3f6176d737fba6a3b6ba3d8/core/variable_map.js#L142)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

I think you don't need to call delete--it should get handled by gc once it's removed from the list, and variableModel doesn't have any special disposal code.
</blockquote></details>

You are right. Took it out.

---

*[core/variable_map.js, line 153 at r1](https://reviewable.io:443/reviews/google/blockly/1079#-KjFN727c16Ee6QrwXP1-r1-153:-KjFPoskjpz-XgAPRfVa:b-896fix) ([raw file](https://github.com/google/blockly/blob/23526f44f79dfa75a3f6176d737fba6a3b6ba3d8/core/variable_map.js#L153)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

the variable with the given name, or null if it was not found.
</blockquote></details>

Done.

---

*[core/variable_map.js, line 188 at r1](https://reviewable.io:443/reviews/google/blockly/1079#-KjFN727c16Ee6QrwXP1-r1-188:-KjFPzd3bE4B31qym0xU:b-896fix) ([raw file](https://github.com/google/blockly/blob/23526f44f79dfa75a3f6176d737fba6a3b6ba3d8/core/variable_map.js#L188)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

"Get a list containing all of the variables of a specified type.  If type is null, ..."
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1079)*

**Comment by marisaleung** [comment]
All tests that just pass through workspace to variable_map functions have been put into a separate file called "variable_map_tests.js".

**Comment by rachel-fenichel** [comment]
:lgtm: if you can get blockly_uncompressed to behave.

---

Reviewed 1 of 4 files at r1, 4 of 4 files at r2.
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1079#-:-KjFiG71bpkW0PIWp7nH:bk57u9c)*

**Pull request merged**